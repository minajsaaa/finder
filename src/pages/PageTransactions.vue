<template lang="pug">
  div
    app-header
    div(v-if="!tx.txsLoaded && tx.txsLoading")
      app-loading
    div(class="txs-container" v-else-if="tx.txsLoaded && !tx.txsLoading && !isEmpty(txs)")
      h2 Transactions
        span for Block {{ `#${$route.params.block}` }}
        p
          span.total A total of {{ txs.length }} transactions found
          app-page(:totalRow="txs.length", v-if="txs.length > 10", v-on:page-change="pageChange")
      ul.tx-table
        li.title
          ul.row
            li
              p.txhash {{ `TxHash` }}
            li
              p.type {{ `Type`}}
            li
              p.txfee {{ `Fee`}}
            li
              p.block {{ `Height` }}
            li
              p.time {{ `Time` }}
        li(v-for="(tx, index) in txs")
          ul.row(v-if="index >= startIndex && index < endIndex")
            li
              router-link.txhash(:to="{ name: 'tx', params: { hash: tx.txhash }}") {{ tx.txhash }}
            li.type
              div {{ tx.tx.type }}
              span(v-if="tx.tx.value.msg.length > 1") {{ `+ ${tx.tx.value.msg.length - 1}` }}
            li
              p.txfee {{ tx.tx.value.fee.amount ? `${rebaseAsset(tx.tx.value.fee.amount[0].amount)} LUNA` : `Null` }}
            li
              router-link.block(:to="{ name: 'block', params: { block: tx.height }}") {{ tx.height }}
            li
              p.time {{ fromNow(blockTime) }}

      app-page(:totalRow="txs.length", v-if="txs.length > 10", v-on:page-change="pageChange")

    template(v-else-if="tx.error && !tx.txsLoading && tx.txsLoaded")
      app-not-found
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import { isEmpty } from "lodash";
import AppHeader from "../components/AppHeader";
import AppPage from "../components/AppPage";
import AppNotFound from "../components/AppNotFound";
import AppLoading from "../components/AppLoading";
import { txToHash, fromNow } from "../scripts/utility";
import { rebaseAsset } from "../scripts/num";

export default {
  beforeCreate: function() {
    document.body.className = "page";
  },
  name: "page-Txs",
  data: () => ({
    startIndex: 0,
    endIndex: 10
  }),
  components: {
    AppHeader,
    AppPage,
    AppNotFound,
    AppLoading
  },
  computed: {
    ...mapGetters(["tx", "block"]),
    txs() {
      const blockHeight = this.$route.params.block;
      const txs = [];

      if (this.block.blocks[blockHeight]) {
        const blockData = this.block.blocks[blockHeight].block;

        const len = blockData.data.txs.length || 0;
        for (let i = 0; i < len; i++) {
          let hash = txToHash(blockData.data.txs[i]);

          txs.push(this.tx.txs[hash]);
        }
      }

      return txs;
    },
    blockMeta() {
      const blockHeight = this.$route.params.block;

      return this.block.blocks[blockHeight].block_meta;
    },
    blockTime() {
      return this.blockMeta.header.time;
    }
  },
  methods: {
    ...mapActions(["queryTxs", "fetchBlock"]),
    fromNow,
    isEmpty,
    rebaseAsset,
    pageChange({ pageNumber, pageSize }) {
      this.startIndex = pageSize * (pageNumber - 1);
      this.endIndex = pageSize * pageNumber;
    }
  },
  async mounted() {
    const blockHeight = this.$route.params.block;
    if (!this.block.blocks[blockHeight]) {
      await this.fetchBlock(this.$route.params);
    }
    await this.queryTxs([
      this.block.blocks[blockHeight].block,
      this.$route.params
    ]);
  },

  watch: {
    // eslint-disable-next-line
    $route(to, from) {
      this.queryTxs([
        this.block.blocks[this.$route.params.block],
        this.$route.params
      ]);
    }
  }
};
</script>

<style lang="stylus">
@require '../styles/variables'

.txs-container
  width 100%
  max-width 1440px
  margin 0 auto
  padding 40px
  color #2043b5 !important

.txs-container h2
  font-size 24px
  font-weight 500
  margin-bottom 20px

.txs-container h2 span
  font-weight 300
  font-size 23px
  padding-left 6px

.txs-container h2
  font-size 24px
  font-weight 500
  margin-bottom 20px

.txs-container h2 p
  margin-top 20px
  height 40px
  display block
  position relative

.txs-container h2 p span.total
  position: absolute;
  font-weight: 400;
  top: 0;
  left: 0;
  font-size 15px
  height 40px
  line-height 40px

.txs-container .row
  display table
  font-size 14px
  min-height 60px
  border-top solid 1px #d8ddf0
  width 100%
  table-layout fixed
  min-width: 850px

.txs-container .row li
  padding 15px
  display table-cell
  vertical-align middle
  line-height 1

.txs-container .row li:first-child
  padding-left 30px

.txs-container .row li:last-child
  padding-left 30px

.txs-container .row li
  width 13%

.txs-container .row li:first-child
  width 48%

.txs-container .tx-table .title
  border-radius 3px 3px 0 0
  font-weight 500

.txs-container .tx-table .title .row
  border-top 0
  background-color rgba(84, 147, 247, 0.1)

.txs-container .txhash, .block, .time, .type, .txfee
  display inline-block
  text-overflow ellipsis
  white-space nowrap
  overflow hidden
  width 100%

.txs-container .type div
  border-radius: 13px;
  height: 26px;
  overflow: hidden;
  background-color: rgba(40, 69, 174,0.1);
  font-size: 12px;
  font-weight: 400;
  font-style: normal;
  font-stretch: normal;
  line-height: 26px;
  letter-spacing: -0.3px;
  color: rgba(40, 69, 174, 1);
  padding 0 15px
  display inline-block
  vertical-align middle

.txs-container .type span
  display: inline-block;
  font-size: 11px;
  vertical-align: middle;
  margin-left: 3px;

.txs-container .tx-table
  border-radius 5px
  border solid 1px #d8ddf0
  width 100%
  margin-bottom 30px
  background #fff
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  overflow-x: auto;
  -webkit-overflow-scrolling: touch;
  min-width unset

@media screen and (max-width: 1280px)
  .txs-container .row li
    width 15%

  .txs-container .row li:first-child
    width 40%

  .txs-container .type div
    font-size 11px
    padding 0 10px

@media screen and (max-width: 1024px)
  .txs-container .row
    min-height 50px
    font-size 13px

  .txs-container .row li
    width 15%

  .txs-container .row li:first-child
    width 40%

  .txs-container .type div
    font-size 11px
    padding 0 10px

  .txs-container .row li
    padding 10px

  .txs-container .row li:first-child
    padding-left 20px

  .txs-container .row li:last-child
    padding-left 20px

@media screen and (max-width: 767px)
  .txs-container
    padding 50px 15px

  .txs-container h2
    padding 40px 0 30px
    margin 0

  .txs-container h2 span
    display block
    padding 0

  .txs-container h2 p
    height auto

  .txs-container h2 p span.total
    position static
    display block
    padding 0
    top: unset;
    left: unset;
    font-size: 14px;
    height: auto;
    line-height: 1.3;

  div.v-pagination > ul
    margin-top 50px
    position relative

  div.v-pagination > ul > li.v-pagination__list
    position absolute
    left 0
    top 0

  div.v-pagination > ul > li.v-pagination__list span
    position absolute
    top: -15px;
    left: 0;
    font-size: 9px;
    line-height: 12px;

  div.v-pagination > ul > li.v-pagination__list select
    margin 0

  div.v-pagination > ul > li.v-pagination__list:after
    right: unset;
    left: 35px;

  .txs-container .tx-table .title .row
    border-top 0

  .txs-container .row li
    width 18%

  .txs-container .row li:nth-child(2)
    width 20%

  .txs-container .row li:first-child
    width 26%
</style>
