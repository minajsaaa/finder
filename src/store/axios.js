import axios from "axios";

const instance = axios.create({
  timeout: 15000
});
export default instance;
