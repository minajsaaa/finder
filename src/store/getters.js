export const block = state => state.block;
export const config = state => state.config;
export const tx = state => state.tx;
export const account = state => state.account;
