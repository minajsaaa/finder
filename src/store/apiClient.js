import axios from "axios";
import { getNetwork } from "../networks";

const limit = 1000;

function lcdUrl(network) {
  return getNetwork(network).lcd;
}

const instance = axios.create({
  timeout: 15000
});

const getBlock = (network, blockHeight, requester = instance) => {
  let url = `${lcdUrl(network)}/blocks/${blockHeight}`;
  return requester.get(url);
};

const getTx = (network, hash, requester = instance) => {
  let url = `${lcdUrl(network)}/txs/${hash.toUpperCase()}`;
  return requester.get(url);
};

const getAccount = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/auth/accounts/${address}`;
  return requester.get(url);
};

const getDelegations = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/staking/delegators/${address}/delegations`;
  return requester.get(url);
};

const getRewards = (
  network,
  delegatorAddress,
  validatorAddress,
  requester = instance
) => {
  let url = `${lcdUrl(
    network
  )}/distribution/delegators/${delegatorAddress}/rewards/${validatorAddress}`;
  return requester.get(url);
};

const getUndelegations = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/staking/delegators/${address}/unbonding_delegations`;
  return requester.get(url);
};

const getSenderTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/txs?sender=${address}&limit=${limit}`;
  return requester.get(url);
};
const getRecipientTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(network)}/txs?recipient=${address}&limit=${limit}`;
  return requester.get(url);
};
const getSwapTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=swap&trader=${address}&limit=${limit}`;
  return requester.get(url);
};
const getProposerTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=submit_proposal&proposer=${address}&limit=${limit}`;
  return requester.get(url);
};
const getDepositorTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/txs?action=deposit&depositor=${address}&limit=${limit}`;
  return requester.get(url);
};
const getDelegatorsTxs = (network, address, requester = instance) => {
  let url = `${lcdUrl(
    network
  )}/staking/delegators/${address}/txs?limit=${limit}`;
  return requester.get(url);
};

export default {
  getBlock,
  getTx,
  getAccount,
  getDelegations,
  getRewards,
  getUndelegations,
  getSenderTxs,
  getRecipientTxs,
  getSwapTxs,
  getProposerTxs,
  getDepositorTxs,
  getDelegatorsTxs
};
