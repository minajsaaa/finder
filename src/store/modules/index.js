import apiClient from "../apiClient";

const files = require.context(".", false, /\.js$/);
let modules = {};

files.keys().forEach(key => {
  if (key === "./index.js") return;
  modules[key.replace(/(\.\/|\.js)/g, "")] = files(key).default(apiClient);
});

export default modules;
