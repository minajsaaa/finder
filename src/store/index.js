import Vue from "vue";
import Vuex from "vuex";
import VueAnalytics from "vue-analytics";
// import * as actions from "./actions";
import * as getters from "./getters";
import modules from "./modules";
import * as Sentry from "@sentry/browser";
import * as Integrations from "@sentry/integrations";

Vue.use(VueAnalytics, {
  id: "UA-122601594-2"
});

Sentry.init({
  dsn: "https://c2041cd0a20e4c7cb6430f1472df397e@sentry.io/1463310",
  integrations: [
    new Integrations.Vue({
      Vue,
      attachProps: true
    })
  ]
});

Vue.use(Vuex);

export default new Vuex.Store({
  getters,
  modules,
  strict: false
});
