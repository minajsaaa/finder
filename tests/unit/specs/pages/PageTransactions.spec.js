import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import * as getters from "@/store/getters";
import PageTransactions from "@/pages/PageTransactions";
import mockState from "../store/mockState";

describe(`PageTransactions`, () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  let wrapper, $store, $route, mockStore;
  const defaultBlock = 719
  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        block: {},
        account: {},
        tx: {}
      }
    };

    $route = {};

    wrapper = shallowMount(PageTransactions, {
      localVue,
      mocks: {
        $store,
        $route
      }
    });
  });

  describe(`Snapshot`, () => {
    beforeEach(() => {
      mockStore = new Vuex.Store({
        state: mockState,
        getters,
        actions: {
          fetchBlock: jest.fn(),
          queryTxs: jest.fn()
        }
      });
    });

    it(`Snapshot: with defaultBlock`, () => {
      wrapper = shallowMount(PageTransactions, {
        localVue,
        store: mockStore,
        mocks: {
          $route: {
            params: {
              block: defaultBlock
            }
          }
        },
        stubs: [`router-link`]
      });
      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });
});
