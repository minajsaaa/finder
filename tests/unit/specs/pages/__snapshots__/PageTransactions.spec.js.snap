// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTransactions Snapshot Snapshot: with defaultBlock 1`] = `
<div>
  <app-header-stub />
  <div
    class="txs-container"
  >
    <h2>
      Transactions
      <span>
        for Block #719
      </span>
      <p>
        <span
          class="total"
        >
          A total of 1 transactions found
        </span>
        <!---->
      </p>
    </h2>
    <ul
      class="tx-table"
    >
      <li
        class="title"
      >
        <ul
          class="row"
        >
          <li>
            <p
              class="txhash"
            >
              TxHash
            </p>
          </li>
          <li>
            <p
              class="type"
            >
              Type
            </p>
          </li>
          <li>
            <p
              class="txfee"
            >
              Fee
            </p>
          </li>
          <li>
            <p
              class="block"
            >
              Height
            </p>
          </li>
          <li>
            <p
              class="time"
            >
              Time
            </p>
          </li>
        </ul>
      </li>
      <li>
        <ul
          class="row"
        >
          <li>
            <router-link-stub
              class="txhash"
              to="[object Object]"
            >
              B463C2868822FD5ABCAAF67FC579691E7060B83139E94CA589E0F56FA10F04D2
            </router-link-stub>
          </li>
          <li
            class="type"
          >
            <div>
              auth/StdTx
            </div>
            <!---->
          </li>
          <li>
            <p
              class="txfee"
            >
              0.003000 LUNA
            </p>
          </li>
          <li>
            <router-link-stub
              class="block"
              to="[object Object]"
            >
              719
            </router-link-stub>
          </li>
          <li>
            <p
              class="time"
            >
              23 days ago
            </p>
          </li>
        </ul>
      </li>
    </ul>
    <!---->
  </div>
</div>
`;
